### PR TITLE
WIP: Plan for CI

### PR DESCRIPTION
**Pull request by jamesmunns**
Hey @nerdyvaishali and @sekineh, here is an initial skeleton, please feel free to add to this PR (you can just commit to the `add-plan` branch, and we can review as we go :)

### CODE REVIEW

**Comment by sekineh** [comment]
I got it, I'll checkout `add-plan` branch.

File: ci-plan.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# Plan for adding embedded CI tests to `rust-lang/rust`
+
+* Goal: we will add target `thumbv7m-none-eabi` in stable Rust 2018 Edition
```
Comment (jamesmunns) [inline ci-plan.md]: We will want to test all four of the Cortex-M targets:

```
thumbv6m-none-eabi - Bare Cortex-M0, M0+, M1
thumbv7em-none-eabi - Bare Cortex-M4, M7
thumbv7em-none-eabihf - Bare Cortex-M4F, M7F, FPU, hardfloat
thumbv7m-none-eabi - Bare Cortex-M3
```

Comment (jamesmunns) [reply ci-plan.md]: Also we will want to test `msp430-none-elf` eventually, but we will focus on the `thumb*` targets first

File: ci-plan.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  * `./x.py dist` for `thumbv7m-none-eabi` is already included in `dist-various-1` docker image; this is why we can do `rustup target add` for the target now.
+
+## Tests to add
+
+> TODO: Add more detail
+
+* Make sure known-good embedded projects still compile and link
+  * [???] We could use `cargotest` infra or alike.
+  * Crate `cortex-m` is a good start because it already compiles on beta now.
+  * `stm32f103xx-hal` includes `cortex-m` in dependency, so we'll use it later.
```
Comment (jamesmunns) [inline ci-plan.md]: :+1:, I agree `cortex-m` is a good place to start for now as it is supported by beta

For more candidates, I would suggest the `examples` folder: https://github.com/japaric/stm32f103xx-hal/tree/master/examples

File: ci-plan.md (hunk 1 above)
Comment (sekineh) [reply ci-plan.md]: Very important point. Now our target became clearer.

File: ci-plan.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  * [???] Check within `cargotest`?  Are there better place?
+
+* (Stage two): Ensure that simple embedded examples run in QEMU
+
+## Plan for integration
+
+> TODO, how/where do we add tests?
+
+* Create a pull request with 'minimal' test for `thumbv7m-none-eabi`
+  * [???] We will create a new docker file for testing: `thumbv7m-none-eabi/Dockerfile`
```
Comment (sekineh) [inline ci-plan.md]: The alternative is to use existing docker image and just do `test && dist` pattern like below. I think it's logically correct for a CI purpose. The only and the biggest drawback is that it takes much more execution time.  We could just do this, and if execution time is problematic, then split up the docker image into several ones.

```
$ rg 'x.py|TARGETS|HOST' src/ci/docker/
(snip)
src/ci/docker/dist-x86_64-musl\Dockerfile
45:      python2.7 ../x.py test --target x86_64-unknown-linux-musl && \
46:      python2.7 ../x.py dist --target x86_64-unknown-linux-musl
(snip)
src/ci/docker/dist-i586-gnu-i586-i686-musl\Dockerfile
49:ENV TARGETS=i586-unknown-linux-gnu,i686-unknown-linux-musl
52:      python2.7 ../x.py test --target $TARGETS && \
53:      python2.7 ../x.py dist --target $TARGETS,i586-unknown-linux-musl
(snip)
```

File: ci-plan.md (hunk 1 above)
Comment (sekineh) [reply ci-plan.md]: For msp430-none-elf, we will need to add ‘dist’ job.

Comment (sekineh) [reply ci-plan.md]: In a big picture, the current status is:
- thumb* targets are already in Tier 2 because it compiles and official binaries available.
- msp430-none-elf is Tier 3 because it has no official binary.
